File: clients/node/src/categories/structs.ts
import { IOneShotClient } from '../types/client.js';
import {
  SolidityStruct,
  SolidityStructParamUpdateRequest,
  NewSolidityStructParam,
} from '../types/struct.js';
import {
  structUpdateSchema,
  addStructParamSchema,
  updateStructParamsSchema,
  removeStructParamSchema,
} from '../validation/struct.js';

export class Structs {
  constructor(private client: IOneShotClient) {}

  /**
   * Update an existing solidity struct
   * @param structId The ID of the struct to update
   * @param name The new name for the struct
   * @returns Promise<SolidityStruct>
   * @throws {ZodError} If the name is invalid
   */
  async update(structId: string, name: string): Promise<SolidityStruct> {
    // Validate the update parameters
    const validatedData = structUpdateSchema.parse({ name });

    return this.client.request<SolidityStruct>('PUT', `/structs/${structId}`, validatedData);
  }

  /**
   * Add a parameter to an existing struct
   * @param businessId The business ID that owns the struct
   * @param structId The ID of the struct to add the parameter to
   * @param param The new parameter to add
   * @returns Promise<SolidityStruct>
   * @throws {ZodError} If the parameter is invalid
   */
  async addParam(
    businessId: string,
    structId: string,
    param: NewSolidityStructParam
  ): Promise<SolidityStruct> {
    // Validate all parameters using the schema
    const validatedParams = addStructParamSchema.parse({
      businessId,
      structId,
      param,
    });

    return this.client.request<SolidityStruct>(
      'POST',
      `/business/${validatedParams.businessId}/structs/${validatedParams.structId}/params`,
      validatedParams.param
    );
  }

  /**
   * Update multiple parameters of a struct
   * @param businessId The business ID that owns the struct
   * @param structId The ID of the struct to update
   * @param updates Array of parameter updates
   * @returns Promise<SolidityStruct>
   * @throws {ZodError} If any of the updates are invalid
   */
  async updateParams(
    businessId: string,
    structId: string,
    updates: SolidityStructParamUpdateRequest[]
  ): Promise<SolidityStruct> {
    // Validate all parameters using the schema
    const validatedParams = updateStructParamsSchema.parse({
      businessId,
      structId,
      updates,
    });

    return this.client.request<SolidityStruct>(
      'PUT',
      `/business/${validatedParams.businessId}/structs/${validatedParams.structId}/params`,
      { updates: validatedParams.updates }
    );
  }

  /**
   * Remove a parameter from a struct
   * @param structId The ID of the struct
   * @param structParamId The ID of the parameter to remove
   * @returns Promise<SolidityStruct>
   * @throws {ZodError} If the parameter ID is invalid
   */
  async removeParam(structId: string, structParamId: string): Promise<SolidityStruct> {
    // Validate all parameters using the schema
    const validatedParams = removeStructParamSchema.parse({
      structId,
      structParamId,
    });

    return this.client.request<SolidityStruct>(
      'DELETE',
      `/structs/${validatedParams.structId}/params/${validatedParams.structParamId}`
    );
  }
}

File: clients/go/pkg/client/structs.go
package client

import (
	"context"

	swagger "github.com/1shotapi/go-client/internal/generated"
)

// Structs handles all solidity struct-related operations
type Structs struct {
	api        *swagger.SolidityStructsApiService
	businessId string
}

// Update updates an existing solidity struct
func (s *Structs) Update(ctx context.Context, structId string, name string) (*swagger.SolidityStruct, error) {
	body := swagger.StructsStructIdBody{Name: name}
	resp, _, err := s.api.StructsStructIdPut(ctx, body, structId)
	if err != nil {
		return nil, err
	}
	return &resp, nil
}

// AddParam adds a parameter to an existing struct
func (s *Structs) AddParam(ctx context.Context, businessId, structId string, param swagger.NewSolidityStructParam) (*swagger.SolidityStruct, error) {
	resp, _, err := s.api.BusinessBusinessIdStructsStructIdParamsPost(ctx, param, businessId, structId)
	if err != nil {
		return nil, err
	}
	return &resp, nil
}

// UpdateParams updates multiple parameters of a struct
func (s *Structs) UpdateParams(ctx context.Context, businessId, structId string, updates []swagger.SolidityStructParamUpdate) (*swagger.SolidityStruct, error) {
	converted := make([]swagger.AllOfstructIdParamsBodyUpdatesItems, len(updates))
	for i, u := range updates {
		var typeStruct *swagger.Object
		if u.TypeStruct != nil {
			obj := swagger.Object{
				"name":   u.TypeStruct.Name,
				"params": u.TypeStruct.Params,
			}
			typeStruct = &obj
		}
		converted[i] = swagger.AllOfstructIdParamsBodyUpdatesItems{
			Id:           u.TypeStructId, // You may need to map the correct ID field here
			Name:         u.Name,
			Description:  u.Description,
			Type_:        u.Type_,
			Index:        u.Index,
			Value:        u.Value,
			TypeSize:     u.TypeSize,
			TypeSize2:    u.TypeSize2,
			IsArray:      u.IsArray,
			ArraySize:    u.ArraySize,
			TypeStructId: u.TypeStructId,
			TypeStruct:   typeStruct,
		}
	}
	body := swagger.StructIdParamsBody{Updates: converted}
	resp, _, err := s.api.BusinessBusinessIdStructsStructIdParamsPut(ctx, body, businessId, structId)
	if err != nil {
		return nil, err
	}
	return &resp, nil
}

// RemoveParam removes a parameter from a struct
func (s *Structs) RemoveParam(ctx context.Context, structId, structParamId string) (*swagger.SolidityStruct, error) {
	resp, _, err := s.api.StructsStructIdParamsStructParamIdDelete(ctx, structId, structParamId)
	if err != nil {
		return nil, err
	}
	return &resp, nil
}
